### PR TITLE
[mixed] feat: Add pretrain for roberta

### PR DESCRIPTION
**Pull request by soeque1**
- Add config
- Add src/roberta/roberta_dataset_builder.py
- Add src/roberta/roberta_runnner.py

Refs: #18

HOW TO:

[WIP] 성능 확인 중

```sh
python tokenizer.py tokenizer/model=roberta
python serialize.py serialize/model=roberta data.tokenizer_dir=tokenizers/roberta
python pretrain.py pretrain/runner=roberta-base +serialize/model=roberta data.tokenizer_dir=tokenizers/roberta trainer.gpus=8 data.per_device_batch_size=24 trainer.accelerator=ddp
```

### CODE REVIEW

**Comment by aisolab** [comment]
현 시점으로서는 저는 LGTM 입니다.

**Comment by bzantium** [comment]
vocab size를 default로 50257로 해주셨는데 그러면 configs/tokenizer/model/roberta.yaml 에서도 수정부탁드립니다!

**Comment by bzantium** [comment]
[cls] [sep] token을 추가하더라도 보통 512로 하는게 맞지 않나요?

**Comment by monologg** [comment]
@bzantium 

https://github.com/pytorch/fairseq/issues/1187

기록차원에서 링크 남겨 놓겠습니다:)

File: src/roberta/roberta_dataset_builder.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+from typing import Dict, List, Union
+
+import datasets
+import transformers
+
+from ..dataset_builder import DatasetBuilder
+from ..utils import list_dict_to_dict_list
+
+
+class RoBERTaDatasetBuilder(DatasetBuilder):
```
Comment (monologg) [inline src/roberta/roberta_dataset_builder.py:10]: @soeque1 

마이너한 부분이긴 한데, `RoBERTa` 명칭을 `Roberta` 로 통일하는 것에 대해서 어떻게 생각하시나요?

이게 `transformers` 가 `RobertaForXXX` 로 사용하다보니, 이런 부분이 다르면 애매한 경우들이 종종 있더라고요ㅠㅠ

File: src/roberta/roberta_dataset_builder.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    def make_examples(self, batch: Dict) -> Dict:
+        examples = []
+        docs = self.tokenizer(
+            batch[self.key], padding=False, add_special_tokens=False, truncation=False, return_attention_mask=False
+        )["input_ids"]
+
+        for doc in docs:
+            doc = [self.tokenizer.cls_token_id] + doc
+            doc += [self.tokenizer.sep_token_id]
```
Comment (monologg) [inline src/roberta/roberta_dataset_builder.py:42]: 민호님께서 얘기해주신대로 `_create_example` 안에서 처리해주는 게 맞는것 같네요!

File: src/roberta/roberta_dataset_builder.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    def _add_document(self, input_ids: List[int]) -> Union[transformers.BatchEncoding, None]:
+        self.buffer += input_ids
+        if len(self.buffer) >= self.max_length:
+            ex = self._create_example()
+            return ex
+        return None
+
+    def _create_example(self) -> transformers.BatchEncoding:
+        chunk = self.buffer[: self.max_length]
```
Comment (monologg) [inline src/roberta/roberta_dataset_builder.py:56]: Q. 지금 구조에서는 context가 다른 문서가 이어붙여지는 건 상관없이, 무조건 full로 이어붙이는 방식인거죠?

File: configs/pretrain/runner/roberta-small.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  resid_pdrop: 0.1
+  embd_pdrop: 0.1
+  attn_pdrop: 0.1
+  layer_norm_epsilon: 1e-5
+  initializer_range: 0.02
+  scale_attn_weights: true
+  gradient_checkpointing: false
+  return_dict: true
+  bos_token_id: 0
+  eos_token_id: 0
```
Comment (monologg) [inline configs/pretrain/runner/roberta-small.yaml:22]: `eos_token_id: 2` 로 바꿔야 할 것 같습니다!

Comment (monologg) [reply configs/pretrain/runner/roberta-small.yaml:22]: https://github.com/lassl/lassl/blob/main/configs/tokenizer/model/roberta.yaml#L18

지금 보니까 special token 순서가 원래 `roberta-base`와 상이하네요

Comment (monologg) [reply configs/pretrain/runner/roberta-small.yaml:22]: https://github.com/lassl/lassl/issues/28

File: src/roberta/roberta_dataset_builder.py (hunk 1 above)
Comment (soeque1) [reply src/roberta/roberta_dataset_builder.py:10]: @monologg 좋습니다!!

File: src/roberta/roberta_dataset_builder.py (hunk 3 above)
Comment (soeque1) [reply src/roberta/roberta_dataset_builder.py:56]: 넵~

File: configs/pretrain/runner/roberta-small.yaml (hunk 4 above)
Comment (soeque1) [reply configs/pretrain/runner/roberta-small.yaml:22]: 버그입니다 ㅋㅋ gpt2 복붙하다가 놓쳤네요 ~ ㅠ

File: src/roberta/roberta_dataset_builder.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+        example = self.tokenizer.prepare_for_model(
+            chunk, padding=False, truncation=False, return_attention_mask=True, return_token_type_ids=False,
+            add_special_tokens=False
```
Comment (bzantium) [inline src/roberta/roberta_dataset_builder.py:60]: 대화 나눴던 것처럼 special_tokens를 뱉어주는 것으로 변경 요청드립니다!

File: src/roberta/roberta_dataset_builder.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        tokenizer: transformers.PreTrainedTokenizerFast,
+        max_length: int,
+        key: str,
+        reset_position_ids=False,
+    ) -> None:
+        self.features = datasets.Features(
+            {
+                "input_ids": datasets.Sequence(datasets.Value("int32")),
+                "attention_mask": datasets.Sequence(datasets.Value("uint8")),
+                "position_ids": datasets.Sequence(datasets.Value("int32")),
```
Comment (bzantium) [inline src/roberta/roberta_dataset_builder.py:24]: 이 부분도 "special_tokens" 추가 요망!

Comment (aisolab) [reply src/roberta/roberta_dataset_builder.py:24]: @bzantium mlm계열의 경우는 serialize 단계에서 `special_tokens_mask`를 받고, runner에서 사용하는 collator가 `[CLS]`와 `[SEP]`는 `special_tokens_mask`정보를 참고하여, `[CLS]`, `[SEP]`에 해당하는 걸 `[MASK]`하지않는 형태가 되겠죠?

Comment (bzantium) [reply src/roberta/roberta_dataset_builder.py:24]: 네 맞습니다! 아래 보시면 해당 부분 prob을 0을 줘서 masking 못하게 하네요 :)
https://github.com/huggingface/transformers/blob/534f6eb9f1f8910a4912ccccd79f1f974c625168/examples/research_projects/performer/run_mlm_performer.py#L247

File: src/roberta/roberta_dataset_builder.py (hunk 5 above)
Comment (soeque1) [reply src/roberta/roberta_dataset_builder.py]: 네 수정했습니당

File: src/roberta/roberta_dataset_builder.py (hunk 6 above)
Comment (soeque1) [reply src/roberta/roberta_dataset_builder.py:24]: 네 ~ 수정했습니다

File: src/roberta/roberta_dataset_builder.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    def _add_document(self, input_ids: List[int]) -> Union[transformers.BatchEncoding, None]:
+        self.buffer += input_ids
+        if len(self.buffer) >= self.max_length - 1:
+            ex = self._create_example()
+            return ex
+        return None
+
+    def _create_example(self) -> transformers.BatchEncoding:
+        chunk = self.buffer[: self.max_length - 1]
+        chunk = [self.tokenizer.cls_token_id] + chunk
```
Comment (bzantium) [inline src/roberta/roberta_dataset_builder.py:57]:
```suggestion
        chunk.insert(0, self.tokenizer.cls_token_id)
```
로 하는 것과 연산 차이가 없을까요?

Comment (soeque1) [reply src/roberta/roberta_dataset_builder.py]: 기존 방식은 리스트를 각각 만들어서 합치는 방식이라 더 개선 될 것 같아요.
실험도 그렇구요

```
import time

def main():
    test_vec = list(range(0, 1000000))
    s = time.time()
    test_vec = [-1] + test_vec
    print(time.time() - s)

    test_vec = list(range(0, 1000000))
    s = time.time()
    test_vec.insert(0, -1)
    print(time.time() - s)


if __name__ == '__main__':
    main()
```

**Review by bzantium — APPROVED** [review]
LGTM! 고생하셨습니다 :)

**Review by Ykmoon — APPROVED** [review]

**Review by monologg — APPROVED** [review]
LGTM!

**Comment by soeque1** [comment]
close #18
